ratelimit/leakybucket.go

```diff
@@ -0,0 +1,118 @@
+package ratelimit
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+
+	"github.com/zalando/skipper/metrics"
+	"github.com/zalando/skipper/net"
+)
+
+// The leaky bucket is an algorithm based on an analogy of how a bucket with a constant leak will overflow if either
+// the average rate at which water is poured in exceeds the rate at which the bucket leaks or if more water than
+// the capacity of the bucket is poured in all at once.
+// See https://en.wikipedia.org/wiki/Leaky_bucket
+type ClusterLeakyBucket struct {
+	capacity    int
+	emission    time.Duration
+	labelPrefix string
+	script      *net.RedisScript
+	ringClient  *net.RedisRingClient
+	metrics     metrics.Metrics
+	now         func() time.Time
+}
+
+const (
+	leakyBucketMetricPrefix  = "leakybucket.redis."
+	leakyBucketMetricLatency = leakyBucketMetricPrefix + "latency"
+	leakyBucketSpanName      = "redis_leakybucket"
+)
+
+// Implements leaky bucket algorithm as a Redis lua script.
+// Redis guarantees that a script is executed in an atomic way:
+// no other script or Redis command will be executed while a script is being executed.
+//
+// Possible optimization: substitute capacity and emission in script source code
+// on script creation in order not to send them over the wire on each call.
+// This way every distinct bucket configuration will get its own script.
+//
+// See https://redis.io/commands/eval
+//
+//go:embed leakybucket.lua
+var leakyBucketScript string
+
+// NewClusterLeakyBucket creates a class of leaky buckets of a given capacity and emission.
+// Emission is the reciprocal of the leak rate and equals the time to leak one unit.
+func NewClusterLeakyBucket(r *Registry, capacity int, emission time.Duration) *ClusterLeakyBucket {
+	return newClusterLeakyBucket(r.redisRing, capacity, emission, time.Now)
+}
+
+func newClusterLeakyBucket(ringClient *net.RedisRingClient, capacity int, emission time.Duration, now func() time.Time) *ClusterLeakyBucket {
+	labelPrefix := fmt.Sprintf("%d-%v-", capacity, emission)
+	return &ClusterLeakyBucket{
+		capacity, emission, labelPrefix,
+		ringClient.NewScript(leakyBucketScript), ringClient, metrics.Default, now,
```

Comment:
ordered assignment?
I think it's too many for this and it's more readable to add the key: value style.

Reply:
key: value would look like this:
```go
	return &ClusterLeakyBucket{
		capacity:    capacity,
		emission:    emission,
		labelPrefix: labelPrefix,
		script:      ringClient.NewScript(leakyBucketScript),
		ringClient:  ringClient,
		metrics:     metrics.Default,
		now:         now,
	}
```
which is stuttering and redundant IMO

Reply:
I think it's much more clean and you can easily lookup from key values to struct member definition.

Reply:
Fixed